src/coreclr/nativeaot/System.Private.CoreLib/src/System/CrashInfo.cs

```diff
@@ -0,0 +1,389 @@
+                return false;
+
+            if (!WriteValue("message", exception.Message, maxMessageSize))
+                return false;
+
+            // Exception type names are not truncated because the full name is important to bucketing and usually not that long
+            if (!WriteValue("type", exception.GetType().ToString()))
+                return false;
+
+            StackFrame[] stackFrames = new StackTrace(exception).GetFrames();
+            if (stackFrames.Length > 0)
+            {
+                if (!OpenValue("stack", '['))
+                    return false;
+
+                int count = 0;
+                foreach (StackFrame frame in stackFrames)
+                {
+                    // Check if the stack frame limit has been hit
+                    if (++count > maxNumberStackFrames)
+                        break;
+
+                    if (!WriteStackFrame(frame, maxMethodNameSize))
+                        return false;
+                }
+
+                CloseValue(']');
+            }
+
+            AggregateException? aggregate = exception as AggregateException;
+            if (aggregate is not null || exception.InnerException is not null)
+            {
+                if (!OpenValue("inner", '['))
+                    return false;
+
+                // Write as many inner exceptions that will fit
+                if (aggregate is not null)
+                {
+                    foreach (Exception ex in aggregate.InnerExceptions)
+                    {
+                        if (!WriteExceptionWithFallback(default, ex, maxMessageSize, maxNumberStackFrames, maxMethodNameSize))
+                            break;
+                    }
+                }
+                else
+                {
+                    WriteExceptionWithFallback(default, exception.InnerException, maxMessageSize, maxNumberStackFrames, maxMethodNameSize);
+                }
+
+                CloseValue(']');
+            }
+            CloseValue('}');
+            return true;
+        }
+
+        /// <summary>
+        /// Write an exception stack frame to the triage buffer
+        /// </summary>
+        /// <param name="frame">the stack frame instance to write</param>
+        /// <param name="maxNameSize">limits the size of the frame type name</param>
+        /// <returns>true - success, false - out of triage buffer space</returns>
+        private bool WriteStackFrame(StackFrame frame, int maxNameSize)
+        {
+            if (!OpenValue(default, '{'))
+                return false;
+
+            nint ip = frame.GetNativeIPAddress();
+            if (!WriteHexValue("ip", (ulong)ip))
+                return false;
+
+            if (!WriteHexValue("offset", frame.GetNativeOffset()))
+                return false;
+
+            string method = DeveloperExperience.GetMethodName(ip, out IntPtr _);
+            if (method != null)
+            {
+                if (!WriteValue("name", method, maxNameSize))
```

Comment:
For names, does it makes sense to truncate from the right?

Reply:
Good point. For type and method names, truncating from the left (somehow) is better.